exchange/reprovide/reprovide.go

```diff
@@ -53,6 +53,11 @@ func (rp *Reprovider) Reprovide(ctx context.Context) error {
 		return fmt.Errorf("Failed to get key chan from blockstore: %s", err)
 	}
 	for k := range keychan {
+		select {
+		case <-ctx.Done():
+			log.Debug("context canceled")
+			return ctx.Err()
+		}
```

Comment:
this is the wrong way to do a context check in this case, something like this is more correct:
```go
for {
    select {
    case <-ctx.Done():
        log.Debug("context canceled")
        return ctx.Err()
    case k, ok := <-keychan:
    ...
```